lib/puppet/http/service/report.rb

```diff
@@ -0,0 +1,43 @@
+class Puppet::HTTP::Service::Report < Puppet::HTTP::Service
+  API = '/puppet/v3'.freeze
+
+  # copied from lib/puppet/indirector/rest.rb
+  EXCLUDED_FORMATS = [:yaml, :b64_zlib_yaml, :dot]
+
+  # puppet major version where JSON is enabled by default
+  MAJOR_VERSION_JSON_DEFAULT = 5
+
+  def initialize(client, server, port)
+    url = build_url(API, server || Puppet[:report_server], port || Puppet[:report_port])
+    super(client, url)
+  end
+
+  def put_report(name, report, environment:, ssl_context: nil)
+    formatter = Puppet::Network::FormatHandler.format_for(Puppet[:preferred_serialization_format])
+
+    # will this always be the same for every call? we can probably freeze the results.
+    model = Puppet::Transaction::Report
+    network_formats = model.supported_formats - EXCLUDED_FORMATS
+    mime_types = network_formats.map { |f| model.get_format(f).mime }
+
+    response = @client.put(
+      with_base_url("/report/#{name}"),
+      headers: { 'ACCEPT' => mime_types.join(', ') },
+      params: { :environment => environment },
+      content_type: formatter.mime,
+      body: formatter.render(report),
+      ssl_context: ssl_context
+    )
+
+    return response.body.to_s if response.success?
+
+    server_version = response[Puppet::Network::HTTP::HEADER_PUPPET_VERSION]
+    if server_version && SemanticPuppet::Version.parse(server_version).major < MAJOR_VERSION_JSON_DEFAULT &&
+        Puppet[:preferred_serialization_format] != 'pson'
+      #TRANSLATORS "pson", "preffered_serialization_format", and "puppetserver" should not be translated
+      raise Puppet::Error.new(_("To submit reports to a server running puppetserver %{server_version}, set preferred_serialization_format to pson") % { server_version: server_version })
```

Comment:
Previously in this scenario, we were emitting a warning and changing `preferred_serialization_format` to pson.

https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/rest.rb#L223-L224

This would be a behavior change.

Reply:
I filed https://tickets.puppetlabs.com/browse/PUP-10147 for the downgrade scenario, so we have some time to update that.